internal/pkg/services/service.go

```diff
@@ -0,0 +1,70 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0 (see LICENSE)
+
+// Contains the several internal services that comprises the cartesi-node and
+// the interfaces and functions used to interact with them
+package services
+
+import (
+	"context"
+	"fmt"
+	"time"
+)
+
+// Defines the interface of a service
+type Service interface {
+	fmt.Stringer
+
+	// Start a service that will run until completion or until the context is
+	// canceled
+	Start(ctx context.Context) error
+}
+
+const DefaultServiceTimeout = 1 * time.Minute
+
+// The Run function serves as a very simple supervisor: it will start all the
+// services provided to it and will run until the first of them finishes. Next
+// it will try to stop the remaining services or timeout if they take too long
+func Run(services []Service) {
+	if len(services) == 0 {
+		panic("there are no services to run")
+	}
+
+	// start services
+	ctx, cancel := context.WithCancel(context.Background())
+	exit := make(chan struct{})
+	for _, service := range services {
+		service := service
+		go func() {
+			if err := service.Start(ctx); err != nil {
+				msg := "main: service '%v' exited with error: %v\n"
+				fmt.Printf(msg, service.String(), err)
+			} else {
+				msg := "main: service '%v' exited successfully\n"
+				fmt.Printf(msg, service.String())
+			}
+			exit <- struct{}{}
+		}()
+	}
+
+	// wait for first service to exit
+	<-exit
+
+	// send stop message to all other services and wait for them to finish
+	// or timeout
+	wait := make(chan struct{})
+	go func() {
+		cancel()
+		for i := 0; i < len(services)-1; i++ {
+			<-exit
+		}
+		wait <- struct{}{}
+	}()
+
+	select {
+	case <-wait:
+		fmt.Println("main: all services exited")
```

Comment:
We should start to use log instead of fmt.Println

Reply:
I agree. I guess we should configure it in the same way we did in the Rust code. I'll update the PR as soon as I figure out how to do it.

Reply:
The `log` package from the standard library is quite simple, it doesn't support different levels or changing the date format. I'll create another issue for this.